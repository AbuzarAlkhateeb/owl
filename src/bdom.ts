/**
 * Block DOM
 *
 * A virtual-dom inspired implementation, but where the basic primitive is a
 * "block" instead of just a html (v)node.
 */

export type BDom = BNode | BMulti | BHtml;

// -----------------------------------------------------------------------------
//  Block
// -----------------------------------------------------------------------------

export abstract class Block {
  el: ChildNode | null | null = null;
  refs?: { [name: string]: HTMLElement };

  mount(parent: HTMLElement | DocumentFragment) {
    const anchor = document.createTextNode("");
    parent.appendChild(anchor);
    this.mountBefore(anchor);
    anchor.remove();
  }

  abstract mountBefore(anchor: ChildNode): void;

  abstract patch(other: Block): void;

  abstract firstChildNode(): ChildNode | null;

  remove() {}

  move(parent: HTMLElement | DocumentFragment) {
    const anchor = document.createTextNode("");
    parent.appendChild(anchor);
    this.moveBefore(anchor);
    anchor.remove();
  }

  moveBefore(anchor: ChildNode): void {
    this.mountBefore(anchor);
  }
}

// -----------------------------------------------------------------------------
//  Html Block
// -----------------------------------------------------------------------------

export class BHtml extends Block {
  html: string;
  content: ChildNode[] = [];
  anchor: ChildNode;
  constructor(html: any) {
    super();
    this.html = String(html);
    this.anchor = document.createTextNode("");
  }

  firstChildNode(): ChildNode | null {
    return this.content[0];
  }

  mountBefore(anchor: ChildNode) {
    this.build();
    anchor.before(this.anchor);
    for (let elem of this.content) {
      this.anchor.before(elem);
    }
  }

  moveBefore(anchor: ChildNode): void {
    anchor.before(this.anchor);
    for (let elem of this.content) {
      this.anchor.before(elem);
    }
  }

  build() {
    const div = document.createElement("div");
    div.innerHTML = this.html;
    this.content = [...div.childNodes];
    this.el = this.content[0];
  }

  remove() {
    for (let elem of this.content) {
      elem.remove();
    }
    this.anchor.remove();
  }

  patch(other: any) {
    for (let elem of this.content) {
      elem.remove();
    }
    this.build();
    for (let elem of this.content) {
      this.anchor.before(elem);
    }
  }

  toString(): string {
    return this.html;
  }
}

// -----------------------------------------------------------------------------
//  Text Block
// -----------------------------------------------------------------------------

export class BText extends Block {
  el: Text;
  text: string;
  constructor(text: string) {
    super();
    this.el = document.createTextNode(text);
    this.text = text;
  }

  firstChildNode(): ChildNode | null {
    return this.el;
  }

  mountBefore(anchor: ChildNode) {
    anchor.before(this.el);
  }

  patch(other: any) {
    if (other.text !== this.text) {
      this.el.textContent = other.el.textContent;
      this.text = other.text;
    }
  }

  toString() {
    return this.el.textContent;
  }
}

// -----------------------------------------------------------------------------
//  Content Block
// -----------------------------------------------------------------------------

export class BNode extends Block {
  static el: ChildNode;
  // el?: HTMLElement | Text;
  children: (BNode | null)[] | null = null;
  anchors?: ChildNode[] | null = null;
  data?: any[] | null = null;
  handlers?: any[] | null = null;

  firstChildNode(): ChildNode | null {
    return this.el;
  }

  toString(): string {
    const div = document.createElement("div");
    this.mount(div);
    return div.innerHTML;
  }

  mountBefore(anchor: ChildNode) {
    this.el = (this.constructor as any).el.cloneNode(true);
    this.build();
    this.update();
    if (this.children) {
      for (let i = 0; i < this.children.length; i++) {
        const child = this.children[i];
        if (child) {
          const anchor = this.anchors![i];
          child.mountBefore(anchor);
        }
      }
    }
    anchor.before(this.el!);
  }

  moveBefore(anchor: ChildNode) {
    anchor.before(this.el!);
  }

  update() {}

  updateClass(elem: HTMLElement, _class: any) {
    switch (typeof _class) {
      case "object":
        for (let k in _class) {
          if (_class[k]) {
            elem.classList.add(k);
          }
        }
        break;
      case "string":
        if (_class) {
          for (let cl of _class.trim().split(/\s+/)) {
            elem.classList.add(cl);
          }
        }
        break;
      default:
        elem.classList.add(_class);
    }
  }

  updateAttr(elem: HTMLElement, attr: string, value: any) {
    if (value !== false) {
      if (value === true) {
        elem.setAttribute(attr, "");
      } else {
        elem.setAttribute(attr, value);
      }
    }
  }

  updateAttrs(elem: HTMLElement, attrs: [string, string] | { [key: string]: string }) {
    if (Array.isArray(attrs)) {
      elem.setAttribute(attrs[0], attrs[1]);
    } else {
      for (let key in attrs) {
        elem.setAttribute(key, attrs[key]);
      }
    }
  }

  updateProp(elem: HTMLElement, prop: string, value: any) {
    (elem as any)[prop] = value;
  }

  setupHandler(el: HTMLElement, index: number) {
    const eventType = this.handlers![index][0];
    el.addEventListener(eventType, (ev) => {
      const info = this.handlers![index];
      const [, callback, ctx] = info;
      if (ctx.__owl__ && !ctx.__owl__.isMounted) {
        return;
      }
      callback(ev);
    });
  }

  protected build() {}

  patch(newTree: any) {
    this.data = newTree.data;
    this.refs = newTree.refs;
    this.update();
    if (this.children) {
      const anchors = this.anchors!;
      const children = this.children;
      const newChildren = newTree.children!;
      for (let i = 0, l = newChildren.length; i < l; i++) {
        const newChild = newChildren[i];
        const child = children[i];
        if (child) {
          if (newChild) {
            child.patch(newChild);
          } else {
            children[i] = null;
            child.remove();
          }
        } else if (newChild) {
          children[i] = newChild;
          newChild.mountBefore(anchors[i]);
        }
      }
    }
  }

  remove() {
    this.el!.remove();
  }
}

export class BStatic extends Block {
  static el: ChildNode;

  firstChildNode(): ChildNode | null {
    return this.el;
  }

  toString(): string {
    const div = document.createElement("div");
    this.mount(div);
    return div.innerHTML;
  }

  mountBefore(anchor: ChildNode) {
    this.el = (this.constructor as any).el.cloneNode(true);
    anchor.before(this.el!);
  }

  moveBefore(anchor: ChildNode) {
    anchor.before(this.el!);
  }

  update() {}

  patch() {}

  remove() {
    this.el!.remove();
  }
}

// -----------------------------------------------------------------------------
//  Multi Block
// -----------------------------------------------------------------------------

export class BMulti extends Block {
  children: (BDom | undefined | null)[];
  anchors?: ChildNode[];

  constructor(n: number) {
    super();
    this.children = new Array(n);
    this.anchors = new Array(n);
  }

  firstChildNode(): ChildNode | null {
    for (let child of this.children) {
      if (child) {
        return child.firstChildNode();
      }
    }
    return null;
  }

  mountBefore(anchor: ChildNode) {
    const children = this.children;
    const anchors = this.anchors;
    for (let i = 0, l = children.length; i < l; i++) {
      let child: any = children[i];
      const childAnchor = document.createTextNode("");
      anchor.before(childAnchor);
      anchors![i] = childAnchor;
      if (child) {
        child.mountBefore(childAnchor);
      }
    }
  }

  moveBefore(anchor: ChildNode) {
    for (let i = 0; i < this.children.length; i++) {
      let child: any = this.children[i];
      const childAnchor = document.createTextNode("");
      anchor.before(childAnchor);
      this.anchors![i] = childAnchor;
      if (child) {
        child.moveBefore(childAnchor);
      }
    }
  }

  patch(newTree: any) {
    const children = this.children;
    const newChildren = newTree.children;
    const anchors = this.anchors!;
    for (let i = 0, l = children.length; i < l; i++) {
      const block = children[i];
      const newBlock = newChildren[i];
      if (block) {
        if (newBlock) {
          block.patch(newBlock);
        } else {
          children[0] = null;
          block.remove();
        }
      } else if (newBlock) {
        children[i] = newBlock;
        newBlock.mountBefore(anchors[i]);
      }
    }
  }

  remove() {
    for (let i = 0; i < this.children.length; i++) {
      this.children[i]!.remove();
      this.anchors![i].remove();
    }
  }

  toString(): string {
    return this.children.map((c) => (c ? c.toString() : "")).join("");
  }
}

// -----------------------------------------------------------------------------
//  Collection Block
// -----------------------------------------------------------------------------

export class BCollection extends Block {
  children: Block[];
  anchor?: ChildNode;
  keys: (string | number)[];

  constructor(n: number) {
    super();
    this.keys = new Array(n);
    this.children = new Array(n);
  }

  firstChildNode(): ChildNode | null {
    return this.children.length ? this.children[0].firstChildNode() : null;
  }

  mountBefore(anchor: ChildNode) {
    const _anchor = document.createTextNode("");
    anchor.before(_anchor);
    this.anchor = _anchor;
    for (let child of this.children) {
      child.mountBefore(_anchor);
    }
  }

  moveBefore(anchor: ChildNode) {
    const _anchor = document.createTextNode("");
    anchor.before(_anchor);
    this.anchor = _anchor;
    for (let child of this.children) {
      child.moveBefore(_anchor);
    }
  }
  patch(other: any) {
    const oldKeys = this.keys;
    const newKeys = other.keys;
    const oldCh = this.children;
    const newCh: Block[] = other.children;
    let oldStartIdx = 0;
    let newStartIdx = 0;
    let oldEndIdx = oldCh.length - 1;
    let newEndIdx = newCh.length - 1;
    let mapping: any = undefined;
    const _anchor = this.anchor!;

    while (oldStartIdx <= oldEndIdx && newStartIdx <= newEndIdx) {
      if (oldCh[oldStartIdx] === null) {
        oldStartIdx++;
      } else if (oldCh[oldEndIdx] === null) {
        oldEndIdx--;
      } else if (oldKeys[oldStartIdx] === newKeys[newStartIdx]) {
        oldCh[oldStartIdx].patch(newCh[newStartIdx]);
        newCh[newStartIdx] = oldCh[oldStartIdx];
        oldStartIdx++;
        newStartIdx++;
      } else if (oldKeys[oldEndIdx] === newKeys[newEndIdx]) {
        oldCh[oldEndIdx].patch(newCh[newEndIdx]);
        newCh[newEndIdx] = oldCh[oldEndIdx];
        oldEndIdx--;
        newEndIdx--;
      } else if (oldKeys[oldStartIdx] === newKeys[newEndIdx]) {
        // bnode moved right
        const elm = oldCh[oldStartIdx];
        elm.patch(newCh[newEndIdx]);
        const nextChild = newCh[newEndIdx + 1];
        const anchor = nextChild ? nextChild.firstChildNode()! : _anchor;
        elm.moveBefore(anchor);
        newCh[newEndIdx] = elm;
        oldStartIdx++;
        newEndIdx--;
      } else if (oldKeys[oldEndIdx] === newKeys[newStartIdx]) {
        // bnode moved left
        const elm = oldCh[oldEndIdx];
        elm.patch(newCh[newStartIdx]);
        const nextChild = oldCh[oldStartIdx];
        const anchor = nextChild ? nextChild.firstChildNode()! : _anchor;
        elm.moveBefore(anchor);
        newCh[newStartIdx] = elm;
        oldEndIdx--;
        newStartIdx++;
      } else {
        mapping = mapping || createMapping(oldKeys, oldStartIdx, oldEndIdx);
        let idxInOld = mapping[newKeys[newStartIdx]];
        if (idxInOld === undefined) {
          // new element
          newCh[newStartIdx].mountBefore(oldCh[oldStartIdx].firstChildNode()!);
          newStartIdx++;
        } else {
          const elmToMove = oldCh[idxInOld];
          elmToMove.moveBefore(oldCh[oldStartIdx].firstChildNode()!);
          elmToMove.patch(newCh[newStartIdx]);
          newCh[newStartIdx] = elmToMove;
          oldCh[idxInOld] = null as any;
          newStartIdx++;
        }
      }
    }
    if (oldStartIdx <= oldEndIdx || newStartIdx <= newEndIdx) {
      if (oldStartIdx > oldEndIdx) {
        const nextChild = newCh[newEndIdx + 1];
        const anchor = nextChild ? nextChild.firstChildNode()! : _anchor;
        for (let i = newStartIdx; i <= newEndIdx; i++) {
          newCh[i].mountBefore(anchor);
        }
      } else {
        for (let i = oldStartIdx; i <= oldEndIdx; i++) {
          let ch = oldCh[i];
          if (ch) {
            ch.remove();
          }
        }
      }
    }
    this.children = newCh;
    this.keys = newKeys;
  }
}

function createMapping(
  oldKeys: any[],
  oldStartIdx: number,
  oldEndIdx: number
): { [key: string]: any } {
  let mapping: any = {};
  for (let i = oldStartIdx; i <= oldEndIdx; i++) {
    mapping[oldKeys[i]] = i;
  }
  return mapping;
}
interface Type<T> extends Function {
  new (...args: any[]): T;
}

export const Blocks: { [key: string]: Type<Block> } = {
  BNode,
  BStatic,
  BMulti,
  BHtml,
  BCollection,
  BText,
};
