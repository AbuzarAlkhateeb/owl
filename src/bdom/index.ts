/**
 * Block DOM
 *
 * A virtual-dom inspired implementation, but where the basic primitive is a
 * "block" instead of just a html (v)node.
 */

import { BCollection } from "./block_collection";
import { BHtml } from "./block_html";
import { BMulti } from "./block_multi";
import { BNode } from "./block_node";
import { BStatic } from "./block_static";
import { BText } from "./block_text";
import { BDispatch } from "./block_dispatch";

export { Block } from "./block";

export const Blocks = {
  BNode,
  BStatic,
  BMulti,
  BHtml,
  BCollection,
  BText,
  BDispatch,
};
