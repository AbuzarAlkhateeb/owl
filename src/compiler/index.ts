import { QWebCompiler, TemplateFunction } from "./compiler";
export { Template, TemplateFunction } from "./compiler";

export function compileTemplate(template: string, name?: string): TemplateFunction {
  const compiler = new QWebCompiler(template, name);
  return compiler.compile();
}
