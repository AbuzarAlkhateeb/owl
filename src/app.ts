import { Blocks } from "./bdom";
import { compileTemplate, Template } from "./compiler/index";
import { callSlot, elem, getValues, owner, scope, toString, withDefault } from "./qweb_utils";
import { Scheduler } from "./scheduler";
import { globalTemplates } from "./tags";

// -----------------------------------------------------------------------------
//  TemplateSet
// -----------------------------------------------------------------------------

export const UTILS = {
  elem,
  toString,
  withDefault,
  call: (name: string) => {
    throw new Error(`Missing template: ${name}`);
  },
  zero: Symbol("zero"),
  scope,
  getValues,
  owner,
  callSlot,
};

export class App {
  rawTemplates: { [name: string]: string } = Object.create(globalTemplates);
  templates: { [name: string]: Template } = {};
  utils: typeof UTILS;
  scheduler = new Scheduler(window.requestAnimationFrame.bind(window));

  constructor() {
    const call = (subTemplate: string, ctx: any, refs: any) => {
      const template = this.getTemplate(subTemplate);
      return template(ctx, refs);
    };

    this.utils = Object.assign({}, UTILS, { call });
  }

  addTemplate(name: string, template: string, allowDuplicate: boolean = false) {
    if (name in this.rawTemplates && !allowDuplicate) {
      throw new Error(`Template ${name} already defined`);
    }
    this.rawTemplates[name] = template;
  }

  getTemplate(name: string): Template {
    if (!(name in this.templates)) {
      const rawTemplate = this.rawTemplates[name];
      if (rawTemplate === undefined) {
        throw new Error(`Missing template: "${name}"`);
      }
      const templateFn = compileTemplate(rawTemplate, name);
      const template = templateFn(Blocks, this.utils);
      this.templates[name] = template;
    }
    return this.templates[name];
  }
}
