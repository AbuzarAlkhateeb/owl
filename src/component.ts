import type { App } from "./app";
import type { BDom } from "./bdom";
import { EventBus } from "./event_bus";
import { Fiber, MountFiber, RootFiber } from "./fibers";

export const enum STATUS {
  CREATED,
  WILLSTARTED, // willstart has been called
  RENDERED, // first render is completed (so, vnode is now defined)
  MOUNTED, // is ready, and in DOM. It has a valid el
  UNMOUNTED, // has a valid el, but is not in DOM
  DESTROYED,
}

export class OwlNode {
  app: App;
  bdom: null | BDom = null;
  component: Component;
  fiber: Fiber | null = null;
  status: STATUS = STATUS.CREATED;
  renderFn: Function;

  constructor(app: App, C: typeof Component, props: any) {
    this.app = app;
    const component = new C(props, app.env, this);
    this.component = component;
    this.renderFn = app.getTemplate(C.template).bind(null, component);
  }
}

export class Component extends EventBus {
  static template: string = "";

  props: any;
  env: any;
  __owl__: OwlNode;

  constructor(props: any, env: any, node: OwlNode) {
    super();
    this.props = props;
    this.env = env;
    this.__owl__ = node;
    this.setup();
  }
  get el() {
    const node = this.__owl__;
    return node.bdom ? node.bdom.el : null;
  }

  setup() {}

  render(): Promise<void> {
    return render(this.__owl__);
  }
}

export function mount<T extends typeof Component>(
  app: App,
  C: T,
  props: any,
  target: HTMLElement
): Promise<InstanceType<T>> {
  const node = new OwlNode(app, C, props);

  const fiber = new MountFiber(node, target);
  node.fiber = fiber;

  app.scheduler.addFiber(fiber);
  // rendering
  fiber.bdom = node.renderFn();
  fiber.counter--;

  return fiber.promise.then(() => node.component);
}

function render(node: OwlNode): Promise<void> {
  const fiber = new RootFiber(node);
  node.fiber = fiber;
  node.app.scheduler.addFiber(fiber);

  fiber.bdom = node.renderFn();
  fiber.counter--;

  return fiber.promise;
}
