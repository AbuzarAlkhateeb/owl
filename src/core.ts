import { Context } from "./context";
import { BDom, Block, Blocks } from "./bdom";
import { ChildFiber, Fiber, MountingFiber, RootFiber } from "./fibers";

// -----------------------------------------------------------------------------
//  Component
// -----------------------------------------------------------------------------

export interface ComponentData {
  bdom: null | BDom;
  render: () => BDom;
  fiber: Fiber | null;
  willStartCB: Function;
  mountedCB: Function;
  isMounted: boolean;
  children: { [key: string]: Component };
  slots?: any;
  context: Context;
}

export class Component {
  static template: string;
  __owl__: ComponentData = currentData;
  props: any;
  env: any = currentEnv;

  constructor(props: any) {
    current = this;
    const __owl__ = currentData;
    __owl__.willStartCB = this.willStart.bind(this);
    __owl__.mountedCB = this.mounted.bind(this);
    this.props = props;
  }

  setup() {}
  async willStart(): Promise<void> {}
  async willUpdateProps(props: any): Promise<void> {}
  mounted() {}

  get el(): ChildNode | null {
    const bdom = this.__owl__.bdom;
    return bdom ? bdom.el : null;
  }

  async render(): Promise<void> {
    const __owl__ = this.__owl__;
    const fiber = new RootFiber(__owl__!);
    __owl__.context.scheduler.addFiber(fiber);
    internalRender(this, fiber);
    await fiber.promise;
  }
}

// -----------------------------------------------------------------------------
//  Component Block
// -----------------------------------------------------------------------------

class BComponent extends Block {
  component: Component;
  handlers?: any[];
  parentClass?: string;

  constructor(name: string, props: any, key: string, ctx: any) {
    super();
    const parentData: ComponentData = ctx.__owl__;
    let component = parentData.children[key];
    if (component) {
      // update
      const fiber = new ChildFiber(component.__owl__, parentData.fiber!);
      const parentFiber = parentData.fiber!;
      parentFiber.child = fiber; // wrong!
      updateAndRender(component, fiber, props);
    } else {
      // new component
      const components = ctx.constructor.components || ctx.components;
      const C = components[name];
      component = prepare(C, props, parentData.context);
      parentData.children[key] = component;
      const fiber = new ChildFiber(component.__owl__, parentData.fiber!);
      const parentFiber = parentData.fiber!;
      parentFiber.child = fiber; // wrong!
      internalRender(component, fiber);
    }
    this.component = component;
  }

  firstChildNode(): ChildNode | null {
    const bdom = this.component.__owl__.bdom;
    return bdom ? bdom.firstChildNode() : null;
  }

  mountBefore(anchor: ChildNode) {
    const bdom = this.component.__owl__!.fiber!.bdom!;
    this.component.__owl__!.bdom = bdom;
    this.component.__owl__!.bdom!.mountBefore(anchor);
    if (this.parentClass) {
      const el = this.firstChildNode();
      if (el instanceof HTMLElement) {
        for (let cl of this.parentClass.trim().split(/\s+/)) {
          el.classList.add(cl);
        }
      }
    }
  }
  patch() {
    this.component.__owl__!.bdom!.patch(this.component.__owl__!.fiber!.bdom);
  }

  remove() {
    const bdom = this.component.__owl__.bdom!;
    bdom.remove();
  }
}

export class BComponentH extends BComponent {
  handlers: any[];
  constructor(handlers: number, name: string, props: any, key: string, ctx: any) {
    super(name, props, key, ctx);
    this.handlers = new Array(handlers);
  }

  mountBefore(anchor: ChildNode) {
    super.mountBefore(anchor);
    this.setupHandlers();
  }

  setupHandlers() {
    for (let i = 0; i < this.handlers.length; i++) {
      const handler = this.handlers[i];
      const eventType = handler[0];
      const el = this.component.el!;
      el.addEventListener(eventType, () => {
        const info = this.handlers![i];
        const [, callback, ctx] = info;
        if (ctx.__owl__ && !ctx.__owl__.isMounted) {
          return;
        }
        callback();
      });
    }
  }
}

async function updateAndRender(component: Component, fiber: ChildFiber, props: any) {
  const componentData = component.__owl__;
  componentData.fiber = fiber;
  await component.willUpdateProps(props);
  component.props = props;
  fiber.bdom = componentData.render();
  fiber.root.counter--;
}

Blocks.BComponent = BComponent;
Blocks.BComponentH = BComponentH;

type Env = any;

interface MountParameters {
  env?: Env;
  target: HTMLElement | DocumentFragment;
  props?: any;
  context?: Context | Component;
}

interface Type<T> extends Function {
  new (...args: any[]): T;
}

let current: Component | null = null;
let currentData: ComponentData;
let currentEnv: any;

export function mount<T extends Type<Component>>(
  C: T | Component,
  params: MountParameters
): Promise<InstanceType<T>>;
export async function mount(C: any, params: MountParameters) {
  if (!(params.target instanceof HTMLElement || params.target instanceof DocumentFragment)) {
    throw new Error("Cannot mount component: the target is not a valid DOM element");
  }
  if (C instanceof Component) {
    // we move component elsewhere
    C.__owl__.bdom!.move(params.target);
    return;
  }
  const { target, props, env, context } = params;
  currentEnv = env || {};
  const componentApp = context
    ? context instanceof Context
      ? context
      : context.__owl__.context
    : new Context();
  const component = prepare(C, props || {}, componentApp);
  const __owl__ = component.__owl__!;
  const fiber = new MountingFiber(__owl__, target);
  __owl__.context.scheduler.addFiber(fiber);
  internalRender(component, fiber);
  await fiber.promise;
  return component;
}

function prepare(C: any, props: any, context: Context): Component {
  let component: Component;
  let template: string = (C as any).template;
  if (!template) {
    throw new Error(`Could not find template for component "${C.name}"`);
  }
  const __owl__: ComponentData = {
    render: null as any,
    bdom: null,
    fiber: null,
    willStartCB: null as any,
    mountedCB: null as any,
    isMounted: false,
    children: {},
    context,
  };
  currentData = __owl__;

  component = new C(props);
  component.setup();
  __owl__.render = context.getTemplate(template).bind(null, component);
  return component;
}

async function internalRender(c: Component, fiber: Fiber) {
  const __owl__ = c.__owl__!;
  __owl__.fiber = fiber;
  await __owl__.willStartCB();
  fiber.bdom = __owl__.render();
  fiber.root.counter--;
}

export function useComponent(): Component {
  return current!;
}

export function useComponentData(): ComponentData {
  return currentData;
}
