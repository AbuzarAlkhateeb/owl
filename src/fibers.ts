import { BDom } from "./bdom";
import { ComponentData } from "./core";

// -----------------------------------------------------------------------------
//  Internal rendering stuff
// -----------------------------------------------------------------------------

class BaseFiber {
  bdom: BDom | null = null;
  error?: Error;
  __owl__: ComponentData;

  child: Fiber | null = null;
  sibling: Fiber | null = null;

  constructor(__owl__: ComponentData) {
    this.__owl__ = __owl__;
  }

  mountComponents() {
    if (this.child) {
      this.child.mountComponents();
    }
    if (this.sibling) {
      this.sibling.mountComponents();
    }
    this.__owl__.mountedCB();
    this.__owl__.isMounted = true;
  }
}

export type Fiber = ChildFiber | RootFiber;

export class ChildFiber extends BaseFiber {
  bdom: BDom | null = null;
  error?: Error;
  root: RootFiber;
  parent: BaseFiber;

  constructor(__owl__: ComponentData, parent: Fiber) {
    super(__owl__);
    this.parent = parent;
    const root = parent.root;
    root.counter++;
    root.childNumber++;
    this.root = root;
  }
}

export class RootFiber extends BaseFiber {
  counter: number = 1;
  childNumber: number = 1;
  root: RootFiber = this;

  resolve!: any;
  reject!: (error: Error) => void;
  promise = new Promise((resolve, reject) => {
    this.resolve = resolve;
    this.reject = reject;
  });

  complete() {
    this.__owl__!.bdom!.patch(this.bdom);
  }
}

export class MountingFiber extends RootFiber {
  target: HTMLElement | DocumentFragment;

  constructor(__owl__: ComponentData, target: HTMLElement | DocumentFragment) {
    super(__owl__);
    this.target = target;
  }
  complete() {
    const __owl__ = this.__owl__!;
    __owl__.bdom! = __owl__.fiber!.bdom!;
    __owl__.bdom!.mount(this.target);
    if (document.body.contains(this.target)) {
      this.mountComponents();
    }
  }
}
