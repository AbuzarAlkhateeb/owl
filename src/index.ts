/**
 * This file is the main file packaged by rollup (see rollup.config.js).  From
 * this file, we export all public owl elements.
 *
 * Note that dynamic values, such as a date or a commit hash are added by rollup
 */
import * as _hooks from "./hooks";
import { useComponent } from "./core";
export { EventBus } from "./event_bus";
export { Context } from "./context";

export { Component, mount, useComponent } from "./core";
export { xml } from "./tags";

export { useState } from "./hooks";

export const hooks = Object.assign(_hooks, { useComponent: useComponent });

export const __info__ = {};
