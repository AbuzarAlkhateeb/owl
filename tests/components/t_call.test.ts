import { Component, mount, useState, xml } from "../../src/index";
import { addTemplate, fromName, makeTestFixture, nextTick, snapshotTemplateCode } from "../helpers";

let fixture: HTMLElement;

beforeEach(() => {
  fixture = makeTestFixture();
});

describe("t-call", () => {
  test("dynamic t-call", async () => {
    class App extends Component {
      static template = xml`
          <t t-call="{{current.template}}">
            owl
          </t>`;
      current = useState({ template: "foo" });
    }
    addTemplate("foo", "<div>foo</div>");
    addTemplate("bar", "bar");

    snapshotTemplateCode(fromName(App.template));
    const app = await mount(App, { target: fixture });
    expect(fixture.innerHTML).toBe("<div>foo</div>");

    app.current.template = "bar";
    await nextTick();
    expect(fixture.innerHTML).toBe("bar");
  });
});
