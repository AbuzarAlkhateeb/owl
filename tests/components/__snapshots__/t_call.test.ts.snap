// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`t-call dynamic t-call 1`] = `
"function anonymous(Blocks, utils
) {
  let {BCollection, BComponent, BComponentH, BHtml, BMulti, BNode, BStatic, BText, BDispatch} = Blocks;
  let {elem, toString, withDefault, call, zero, callSlot, capture} = utils;
  
  return (ctx, parent = ctx) => {
    ctx = Object.create(ctx);
    const b1 = new BText(\` owl \`);
    ctx[zero] = b1;
    const template1 = (ctx['current'].template);
    const b2 = new BDispatch(template1, call(template1, ctx, parent));
    return b2;
  }
}"
`;
